In [1]:
import numpy as np
import pandas as pd
pd.options.display.max_rows = 999

##### I will be using different source for scraping instead of Wiki. See URL below.

In [2]:
tables = pd.read_html('https://www.zipcodesonline.com/2020/06/postal-code-of-toronto-in-2020.html')

In [3]:
tn_df = tables[1]
tn_df.head()

0                1             2                 3
0  SL. NO.    NEIGHBOURHOOD  POSTAL CODES          DISTRICT
1      NaN              NaN           NaN               NaN
2        1         Adelaide           M5H  Downtown Toronto
3        2  Agincourt North           M1V       Scarborough
4        3        Agincourt           M1S       Scarborough

##### Droping  non useful rows 0 and 1.

In [4]:
tn_df2 = tn_df.drop(index =[0,1]).drop(columns=0)

##### Changing columns names.

In [5]:
tn_df3 = tn_df2.reset_index(drop=True)
tn_df3.columns = ['Neighborhood','PostalCode','Borough']

In [6]:
tn_df4 = tn_df3[['PostalCode','Borough','Neighborhood']]
tn_df4

PostalCode           Borough  \
0             M5H  Downtown Toronto   
1             M1V       Scarborough   
2             M1S       Scarborough   
3             M9V         Etobicoke   
4             M8W         Etobicoke   
5             M3H        North York   
6             M5V  Downtown Toronto   
7             M2K        North York   
8             M9V         Etobicoke   
9             M5M        North York   
10            M5E  Downtown Toronto   
11            M1N       Scarborough   
12            M9C         Etobicoke   
13            M4J         East York   
14            M6K      West Toronto   
15            M7Y      East Toronto   
16            M4X  Downtown Toronto   
17            M6E              York   
18            M7R       Mississauga   
19            M1H       Scarborough   
20            M5G  Downtown Toronto   
21            M5T  Downtown Toronto   
22            M6G  Downtown Toronto   
23            M4Y  Downtown Toronto   
24            M1L       Scarborough   
25            M1T       Scarborough   
26            M1M       Scarborough   
27            M1N       Scarborough   
28            M1M       Scarborough   
29            M9B         Etobicoke   
30            M5V  Downtown Toronto   
31            M5L  Downtown Toronto   
32            M4P   Central Toronto   
33            M4S   Central Toronto   
34            M4V   Central Toronto   
35            M6M              York   
36            M5K  Downtown Toronto   
37            M3B       North York,   
38            M1P       Scarborough   
39            M6H      West Toronto   
40    M3M,M3L,M3N        North York   
41            M3H        North York   
42            M3K        North York   
43            M6H      West Toronto   
44            M1K       Scarborough   
45            M4J         East York   
46            M9M        North York   
47            M9C         Etobicoke   
48            M6K      West Toronto   
49            M2J        North York   
50            M5X  Downtown Toronto   
51            M5P   Central Toronto   
52            M5P   Central Toronto   
53            M4V   Central Toronto   
54            M5B  Downtown Toronto   
55            M6B        North York   
56            M1L       Scarborough   
57            M5T  Downtown Toronto   
58            M1E       Scarborough   
59            M5S  Downtown Toronto   
60            M5J  Downtown Toronto   
61            M5V  Downtown Toronto   
62            M5A  Downtown Toronto   
63            M2J        North York   
64            M6P      West Toronto   
65            M1C       Scarborough   
66            M2H        North York   
67            M8V         Etobicoke   
68            M8Y         Etobicoke   
69            M9L        North York   
70            M9A         Etobicoke   
71            M9V         Etobicoke   
72            M9M        North York   
73            M6C              York   
74            M4L      East Toronto   
75            M1K       Scarborough   
76            M5V  Downtown Toronto   
77       M9B, M9A         Etobicoke   
78            M9V         Etobicoke   
79            M6M              York   
80            M1K       Scarborough   
81            M5T  Downtown Toronto   
82            M5V  Downtown Toronto   
83            M8Y         Etobicoke   
84            M9R         Etobicoke   
85            M8Y         Etobicoke   
86            M8Z         Etobicoke   
87            M1V       Scarborough   
88            M1W       Scarborough   
89            M6A        North York   
90            M5M        North York   
91            M6A        North York   
92            M4R   Central Toronto   
93            M4N   Central Toronto   
94            M4G         East York   
95            M6J      West Toronto   
96            M8W         Etobicoke   
97            M1B       Scarborough   
98            M6L        North York   
99            M9C         Etobicoke   
100           M9R         Etobicoke   
101           M9B         Eto

##### Now we need to clean data in column PostalCodes. Three rows have 2+ postal codes togeather. So, we will brake them down. *We will write indexes of rows which we are braking down to be sute to drop them later easily.

In [7]:
drop_indx = []
multy_pc_df = pd.DataFrame(columns = ['PostalCode','Borough','Neighborhood'])
for indx, series in tn_df4.iterrows():
    if len(series.PostalCode) > 3:
        drop_indx.append(series.name)
        postal_codes = series.PostalCode.split(',')
        for pc in postal_codes:
            multy_pc_df = multy_pc_df.append({'PostalCode' : pc,
                                                'Borough' : series.Borough,
                                                'Neighborhood' : series.Neighborhood},
                                               ignore_index=True)
print(drop_indx)
multy_pc_df

[40, 77, 197]


PostalCode     Borough      Neighborhood
0        M3M  North York         Downsview
1        M3L  North York         Downsview
2        M3N  North York         Downsview
3        M9B   Etobicoke  Islington Avenue
4        M9A   Etobicoke  Islington Avenue
5        M2M  North York        Willowdale
6        M2N  North York        Willowdale
7        M2R  North York        Willowdale

##### Now concat new broken DF with main DF where we droping those 3 rows which we have broken down.

In [8]:
tn_df5 = pd.concat([multy_pc_df, tn_df4.drop(drop_indx)])
tn_df5 = tn_df5.reset_index(drop=True)

##### Calculating unique PostalCodes before grouping by Postal Code and Borough to be sure we haven't missed any of them after groupping and function for combining neighborhoods with the same postal code in one row. 

In [9]:
tn_df5.nunique()

PostalCode      103
Borough          11
Neighborhood    201
dtype: int64

##### Combining function. Puts all neighborhoods with the same postal code to the same row. Then we check if all rows has been preserved. We have 103 - the same as qunique values from the main frame.

In [10]:
def combine_n(group):
    (group_label, df) = group
    if df.shape[1] > 1:
        neighborhoods = df.Neighborhood.tolist()
        df.iloc[0, df.columns.get_loc('Neighborhood')] = ', '.join(neighborhoods)
        df2 = df.iloc[0]
        return df2
    else:
        return df.iloc[0]

groups = tn_df5.groupby(by=['PostalCode','Borough'])
data = [combine_n(g) for g in groups]

combined_df = pd.DataFrame(data)
combined_df.shape   

(103, 3)

##### Last touches: Reseting index and text aligning.

In [11]:
comb_toronto_df = combined_df.reset_index(drop=True)
comb_toronto_df.style.set_properties(**{'text-align': 'left'})

In [12]:
comb_toronto_df.to_pickle('toronto-neighborhoods-df.pkl')